lib/associations/has-many.js

```diff
@@ -190,152 +189,120 @@ HasMany.prototype.injectGetter = function(obj) {
 };
 
 HasMany.prototype.injectSetter = function(obj) {
-  var association = this
-    , primaryKeyAttribute = association.target.primaryKeyAttribute;
+  var association = this;
 
   obj[this.accessors.set] = function(newAssociatedObjects, additionalAttributes) {
+    var options = additionalAttributes || {};
     additionalAttributes = additionalAttributes || {};
 
     if (newAssociatedObjects === null) {
       newAssociatedObjects = [];
     } else {
-      newAssociatedObjects = newAssociatedObjects.map(function(newAssociatedObject) {
-        if (!(newAssociatedObject instanceof association.target.Instance)) {
-          var tmpInstance = {};
-          tmpInstance[primaryKeyAttribute] = newAssociatedObject;
-          return association.target.build(tmpInstance, {
-            isNewRecord: false
-          });
-        }
-        return newAssociatedObject;
-      });
+      newAssociatedObjects = association.toInstanceArray(newAssociatedObjects);
     }
 
     var instance = this;
 
-    return instance[association.accessors.get]({
+    return instance[association.accessors.get](_.defaults({
       scope: false,
-      transaction: (additionalAttributes || {}).transaction,
-      logging: (additionalAttributes || {}).logging
-    }).then(function(oldAssociatedObjects) {
-      return new HasManySingleLinked(association, instance).injectSetter(oldAssociatedObjects, newAssociatedObjects, additionalAttributes);
-    });
-  };
-
-  obj[this.accessors.addMultiple] = obj[this.accessors.add] = function(newInstance, options) {
-    // If newInstance is null or undefined, no-op
-    if (!newInstance) return Utils.Promise.resolve();
+      raw: true
```

Comment:
RAWR! 

Just an attempt at making it slightly more performant - we only ever use the data internally